ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -127,15 +169,102 @@ muxChannel pmss mid md w cnt =
         -- We send CBOR encoded messages by encoding them into by ByteString
         -- forwarding them to the 'mux' thread, see 'Desired servicing semantics'.
         --say $ printf "send mid %s mode %s" (show mid) (show md)
+        when (BL.length encoding > maximumMessageSize mid) $
+            throwM $ MuxError MuxTooLargeMessage
+                (printf "Attempting to send a message of size %d on %s %s" (BL.length encoding)
+                        (show mid) (show $ md))
+                callStack
         atomically $ modifyTVar' cnt (+ 1)
         atomically $ putTMVar w encoding
         atomically $ writeTBQueue (tsrQueue pmss) (TLSRDemand mid md (Wanton w))
     recv = do
         -- We receive CBOR encoded messages as ByteStrings (possibly partial) from the
         -- matching ingress queueu. This is the same queue the 'demux' thread writes to.
-        blob <- atomically $ readTBQueue (ingressQueue (dispatchTable pmss) mid md)
-        --say $ printf "recv mid %s mode %s blob len %d" (show mid) (show md) (BL.length blob)
+        blob <- atomically $ do
+            let q = ingressQueue (dispatchTable pmss) mid md
+            blob <- readTVar q
+            if blob == BL.empty
+                then retry
+                else writeTVar q BL.empty >> return blob
+        -- say $ printf "recv mid %s mode %s blob len %d" (show mid) (show md) (BL.length blob)
         if BL.null blob
            then pure Nothing
```

Comment:
Inaccessible code?

```
null :: ByteString -> Bool
null Empty = True
null _     = False

empty :: ByteString
empty = Empty
```

and `blob /= Empty`.

Reply:
Yes, I brought it up on slack last week since I wanted to change `Channel`'s `recv` interface to not return a `Maybe` since with this change it is not possible to use an empty string as an end-of-stream indicator anymore.